app/src/main/java/xyz/block/gosling/features/agent/providers/OpenAIProviderHandler.kt

```diff
@@ -0,0 +1,103 @@
+package xyz.block.gosling.features.agent.providers
+
+import kotlinx.serialization.encodeToString
+import kotlinx.serialization.json.Json
+import org.json.JSONObject
+import xyz.block.gosling.features.agent.*
+import java.lang.reflect.Method
+import java.util.*
+
+/**
+ * OpenAI provider handler for GPT models.
+ * Handles OpenAI-specific API format and tool calling conventions.
+ */
+class OpenAIProviderHandler : LLMProviderHandler {
+    
+    private val jsonFormat = Json {
+        prettyPrint = true
+        encodeDefaults = true
+        ignoreUnknownKeys = true
+    }
+    
+    override fun createToolDefinitions(toolMethods: List<Method>): SerializableToolDefinitions {
+        val toolDefinitions = toolMethods.mapNotNull { method ->
+            val tool = method.getAnnotation(Tool::class.java) ?: return@mapNotNull null
+
+            // Always create a ToolParametersObject, even for tools with no parameters
+            val toolParameters = ToolParametersObject(
+                properties = tool.parameters.associate { param ->
+                    param.name to ToolParameter(
+                        type = param.type,
+                        description = param.description
+                    )
+                },
+                required = tool.parameters
+                    .filter { it.required }
+                    .map { it.name }
+            )
+
+            ToolDefinition(
+                function = ToolFunctionDefinition(
+                    name = tool.name,
+                    description = tool.description,
+                    parameters = toolParameters
+                )
+            )
+        }
+        
+        return SerializableToolDefinitions.OpenAITools(toolDefinitions)
+    }
+    
+    override fun createRequest(
+        modelIdentifier: String,
+        messages: List<Message>,
+        tools: SerializableToolDefinitions,
+        apiKey: String?
+    ): String {
+        val toolDefinitions = when (tools) {
+            is SerializableToolDefinitions.OpenAITools -> tools.definitions
+            else -> emptyList()
+        }
+
+        val openAIRequest = OpenAIRequest(
+            model = modelIdentifier,
+            messages = messages,
+            temperature = if (modelIdentifier != "o3-mini") 0.1 else null,
+            tools = toolDefinitions
+        )
+
+        return jsonFormat.encodeToString(openAIRequest)
```

Comment:
(maybe) me being picky: Is there a reason we declare openAIRequest  but only use it once?
I'm thinking depending on how big the value returned by `OpenAIRequest()` it maybe be a waste of memory?

```suggestion
        return jsonFormat.encodeToString(OpenAIRequest(
            model = modelIdentifier,
            messages = messages,
            temperature = if (modelIdentifier != "o3-mini") 0.1 else null,
            tools = toolDefinitions
        ))
```

Let me know, maybe I'm totally kidding myself

Reply:
Nice! I went ahead and included this in my latest push, thanks @bricefriha :)